dask/operation.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+from typing import Hashable
+
+from dask.base import tokenize
+
+
+class CollectionOperation:
+    """CollectionOperation class
+
+    Encapsulates the state and graph-creation logic for
+    a generic Dask collection.
+    """
+
+    @property
+    def name(self) -> str:
+        """Return the name for this CollectionOperation
+
+        This name should match the string label used in the
+        output keys of a task grah terminating with this operation.
+        """
+        raise NotImplementedError
+
+    @property
+    def dependencies(self) -> set[CollectionOperation]:
+        """Return set of CollectionOperation dependencies"""
+        raise NotImplementedError
+
+    def subgraph(self, keys: list[tuple]) -> tuple:
+        """Return the subgraph and key dependencies for this operation
+
+        NOTE: Since this method returns a mapping between dependencies
+        and required keys, the logic may include recursively fusion.
+
+        Parameters
+        ----------
+        keys : list[tuple]
+            List of required output keys needed from this collection
+
+        Returns
+        -------
+        graph : dict
+            The subgraph for the current operation
+        dependencies : dict[CollectionOperation, list[tuple]]
+            A dictionary mapping ``CollectionOperation`` objects
+            to required keys. This dictionary will be used by the
+            global graph-generation algorithm to determine which
+            operation-key combinations need to be materialized after
+            this operation.
+        """
+        raise NotImplementedError
+
+    def reinitialize(self, new_dependencies: dict, **new_kwargs) -> CollectionOperation:
```

Comment:
Since `CollectionOperation` inheriting classes are going to implement this method the return type should be a bound `TypeVar` (eventually a [`Self`](https://peps.python.org/pep-0673/)); perhaps something like
```python
CollOpT = TypeVar("CollOpT", bound="CollectionOperation")
```